File: services/backend/app/services/explore.py
from __future__ import annotations

from collections.abc import Iterable
from statistics import mean

from app.schemas.explore import (
    BreathingExerciseModule,
    BreathingStep,
    ExploreModule,
    ExploreModulesResponse,
    ExploreModuleType,
    PsychoeducationModule,
    PsychoeducationResource,
    TrendingTopic,
    TrendingTopicsModule,
)
from app.schemas.reports import JourneyReportsResponse
from app.services.feature_flags import FeatureFlagService
from app.services.reports import ReportsService
from app.services.translation import TranslationService


class ExploreService:
    """Compose Explore page modules with feature-flag-aware personalization."""

    def __init__(
        self,
        feature_flags: FeatureFlagService,
        reports_service: ReportsService,
        *,
        translator: TranslationService | None = None,
    ) -> None:
        self._feature_flags = feature_flags
        self._reports_service = reports_service
        self._translator = translator

    async def build_modules(self, *, user_id: str, locale: str = "zh-CN") -> ExploreModulesResponse:
        evaluated: dict[str, bool] = {}
        modules: list[ExploreModule] = []
        reports: JourneyReportsResponse | None = None

        async def ensure_reports() -> JourneyReportsResponse:
            nonlocal reports
            if reports is None:
                reports = await self._reports_service.get_reports(user_id)
            return reports

        breathing_enabled = await self._is_enabled("explore_breathing", subject_id=user_id)
        evaluated["explore_breathing"] = breathing_enabled
        if breathing_enabled:
            module = await self._build_breathing_module(locale=locale)
            module.feature_flag = "explore_breathing"
            modules.append(module)

        psychoeducation_enabled = await self._is_enabled("explore_psychoeducation", subject_id=user_id)
        evaluated["explore_psychoeducation"] = psychoeducation_enabled
        if psychoeducation_enabled:
            payload = await ensure_reports()
            module = await self._build_psychoeducation_module(locale=locale, reports=payload)
            module.feature_flag = "explore_psychoeducation"
            modules.append(module)

        trending_enabled = await self._is_enabled("explore_trending", subject_id=user_id)
        evaluated["explore_trending"] = trending_enabled
        if trending_enabled:
            payload = await ensure_reports()
            module = await self._build_trending_module(locale=locale, reports=payload)
            module.feature_flag = "explore_trending"
            modules.append(module)

        return ExploreModulesResponse(locale=locale, modules=modules, evaluated_flags=evaluated)

    async def _build_breathing_module(
        self,
        *,
        locale: str,
        reports: JourneyReportsResponse | None = None,  # noqa: ARG002
    ) -> BreathingExerciseModule:
        """Return a breathing module translated at runtime."""
        step_data = [
            {
                "label": "Posture",
                "instruction": "Sit or stand tall, relax your shoulders, close your eyes.",
                "duration": 10,
            },
            {
                "label": "Inhale 4 Count",
                "instruction": "Inhale through the nose while counting 1-2-3-4.",
                "duration": 16,
            },
            {
                "label": "Hold 7 Count",
                "instruction": "Hold gently for seven beats without tensing your neck.",
                "duration": 28,
            },
            {
                "label": "Exhale 8 Count",
                "instruction": "Exhale slowly through the mouth, feeling your chest soften.",
                "duration": 32,
            },
        ]
        base_strings = {
            "title": "Guided Breathing Session",
            "description": "A five-minute pace-reset that slows breathing and lowers heart rate.",
            "cadence": "4-7-8 cadence",
            "frequency": "Practice 2-3 rounds before bed or when anxiety spikes.",
            "cta": "Start breathing guide",
        }

        if self._should_translate(locale):
            payload: dict[str, str] = {
                "title": base_strings["title"],
                "description": base_strings["description"],
                "cadence": base_strings["cadence"],
                "frequency": base_strings["frequency"],
                "cta": base_strings["cta"],
            }
            for index, step in enumerate(step_data):
                payload[f"step_label_{index}"] = step["label"]
                payload[f"step_instruction_{index}"] = step["instruction"]

            translated = await self._translate_mapping(payload, locale=locale)
            steps = [
                BreathingStep(
                    label=translated[f"step_label_{index}"],
                    instruction=translated[f"step_instruction_{index}"],
                    duration_seconds=step["duration"],
                )
                for index, step in enumerate(step_data)
            ]
            title = translated["title"]
            description = translated["description"]
            cadence_label = translated["cadence"]
            frequency = translated["frequency"]
            cta_label = translated["cta"]
        else:
            steps = [
                BreathingStep(
                    label=step["label"],
                    instruction=step["instruction"],
                    duration_seconds=step["duration"],
                )
                for step in step_data
            ]
            title = base_strings["title"]
            description = base_strings["description"]
            cadence_label = base_strings["cadence"]
            frequency = base_strings["frequency"]
            cta_label = base_strings["cta"]

        return BreathingExerciseModule(
            id="breathing-reset",
            title=title,
            description=description,
            cadence_label=cadence_label,
            steps=steps,
            recommended_frequency=frequency,
            cta_label=cta_label,
            cta_action="/app/practices/breathing",
        )

    async def _build_psychoeducation_module(
        self,
        *,
        locale: str,
        reports: JourneyReportsResponse,
    ) -> PsychoeducationModule:
        """Return psychoeducation articles aligned with recent themes."""
        themes = self._collect_recent_themes(reports)
        if not themes:
            themes = self._fallback_themes()

        primary_theme = themes[0]

        resources = [
            PsychoeducationResource(
                id="micro-steps",
                title="Break anxiety into micro-actions",
                summary="A three-step framework for interrupting anxious spirals with breathing, logging, and reframing.",
                read_time_minutes=6,
                tags=[primary_theme, "self-awareness"],
            ),
            PsychoeducationResource(
                id="sleep-hygiene",
                title="Rebuild your sleep rhythm",
                summary="Design a gentle wind-down routine using light cues, journaling, and body relaxation.",
                read_time_minutes=5,
                tags=["sleep hygiene", "relaxation"],
            ),
            PsychoeducationResource(
                id="body-scan",
                title="Body scan in 3 minutes",
                summary="Short guided audio to locate tension, soften breathing, and re-anchor attention.",
                read_time_minutes=3,
                tags=["mindfulness"],
                resource_type="audio",
            ),
        ]

        title = "Featured psychoeducation"
        description = "Personalized psychoeducation picks that mirror your recent conversations."
        cta_label = "Open resource hub"

        if self._should_translate(locale):
            mapping: dict[str, str] = {
                "title": title,
                "description": description,
                "cta": cta_label,
            }
            for index, resource in enumerate(resources):
                mapping[f"resource_title_{index}"] = resource.title
                mapping[f"resource_summary_{index}"] = resource.summary

            translated = await self._translate_mapping(mapping, locale=locale)
            translated_resources: list[PsychoeducationResource] = []
            for index, resource in enumerate(resources):
                translated_resources.append(
                    PsychoeducationResource(
                        id=resource.id,
                        title=translated[f"resource_title_{index}"],
                        summary=translated[f"resource_summary_{index}"],
                        read_time_minutes=resource.read_time_minutes,
                        tags=await self._translate_list(resource.tags, locale=locale),
                        resource_type=resource.resource_type,
                        url=resource.url,
                    )
                )
            resources = translated_resources
            title = translated["title"]
            description = translated["description"]
            cta_label = translated["cta"]

        return PsychoeducationModule(
            id="psychoeducation",
            title=title,
            description=description,
            resources=resources,
            cta_label=cta_label,
            cta_action="/app/library",
        )

    async def _build_trending_module(
        self,
        *,
        locale: str,
        reports: JourneyReportsResponse,
    ) -> TrendingTopicsModule:
        topics = self._collect_recent_themes(reports)
        if not topics:
            topics = self._fallback_themes()

        daily_deltas = [report.mood_delta for report in reports.daily if report.mood_delta is not None]
        average_mood = mean(daily_deltas) if daily_deltas else 0.0
        trend = "up" if average_mood > 0.4 else "down" if average_mood < -0.4 else "steady"
        base_momentum = 68 if trend == "up" else 48 if trend == "steady" else 36

        localized_topics: list[TrendingTopic] = []
        topic_names = topics[:3]
        summaries: list[str] = []
        for index, topic in enumerate(topic_names):
            adjustment = max(-12, min(12, int(round(average_mood * 10))))
            momentum = max(20, min(95, base_momentum - index * 6 + adjustment))
            summary = f'Discussions around "{topic}" are gaining traction; pair it with breathing or journaling.'
            summaries.append(summary)
            localized_topics.append(
                TrendingTopic(
                    name=topic,
                    momentum=momentum,
                    trend=trend,  # type: ignore[arg-type]
                    summary=summary,
                )
            )

        insights: list[str] = []
        if reports.daily:
            insights.append(reports.daily[0].spotlight)
        if reports.weekly:
            insights.append(reports.weekly[0].highlights)
        if not insights:
            insights = self._fallback_insights()

        title = "Trending focus areas"
        description = (
            "Based on your latest chats and summaries, these themes deserve extra attention."
        )
        cta_label = "View practice ideas"

        if self._should_translate(locale):
            mapping = {
                "title": title,
                "description": description,
                "cta": cta_label,
            }
            translated = await self._translate_mapping(mapping, locale=locale)
            translated_names = await self._translate_list(topic_names, locale=locale)
            translated_summaries = await self._translate_list(summaries, locale=locale)
            translated_insights = await self._translate_list(insights, locale=locale)

            localized_topics = [
                TrendingTopic(
                    name=translated_names[index],
                    momentum=topic.momentum,
                    trend=topic.trend,
                    summary=translated_summaries[index],
                )
                for index, topic in enumerate(localized_topics)
            ]
            insights = translated_insights
            title = translated["title"]
            description = translated["description"]
            cta_label = translated["cta"]

        return TrendingTopicsModule(
            id="trending-topics",
            title=title,
            description=description,
            topics=localized_topics,
            insights=insights[:3],
            cta_label=cta_label,
            cta_action="/app/trends",
        )

    async def _is_enabled(self, key: str, *, subject_id: str | None) -> bool:
        try:
            evaluation = await self._feature_flags.evaluate_flag(key, subject_id=subject_id)
        except ValueError:
            return True
        return evaluation.enabled

    def _collect_recent_themes(self, reports: JourneyReportsResponse) -> list[str]:
        themes: list[str] = []
        for weekly in reports.weekly:
            for theme in weekly.themes:
                normalized = theme.strip()
                if not normalized or normalized in themes:
                    continue
                themes.append(normalized)
                if len(themes) >= 4:
                    break
            if len(themes) >= 4:
                break
        return themes

    def _fallback_themes(self) -> list[str]:
        return ["Stress management", "Sleep rhythm", "Emotion regulation"]

    def _fallback_insights(self) -> list[str]:
        return [
            "A ten-minute wind-down ritual helps shorten sleep onset time.",
            "Logging triggers plus a 4-7-8 breathing round steadies heart rate within minutes.",
        ]

    def _should_translate(self, locale: str) -> bool:
        if not self._translator:
            return False
        return not self._translator.are_locales_equivalent(locale, "en-US")

    async def _translate_list(self, values: Iterable[str], *, locale: str) -> list[str]:
        items = list(values)
        if not items or not self._should_translate(locale):
            return items
        assert self._translator is not None
        return await self._translator.translate_list(
            items,
            target_locale=locale,
            source_locale="en-US",
        )

    async def _translate_mapping(self, mapping: dict[str, str], *, locale: str) -> dict[str, str]:
        if not mapping:
            return {}
        if not self._should_translate(locale):
            return dict(mapping)
        assert self._translator is not None
        keys = list(mapping.keys())
        values = [mapping[key] for key in keys]
        translated_values = await self._translator.translate_list(
            values,
            target_locale=locale,
            source_locale="en-US",
        )
        return {key: translated_values[index] for index, key in enumerate(keys)}

File: clients/web/src/api/explore.ts
import { getApiBaseUrl, withAuthHeaders } from "./client";
import { asArray, asNumber, asRecord, asString, asStringArray } from "./parsing";
import type {
  BreathingModule,
  ExploreModule,
  ExploreModulesResponse,
  ExploreModuleType,
  PsychoeducationModule,
  PsychoeducationResource,
  TrendingTopic,
  TrendingTopicsModule
} from "./types";

export type ExploreModulesSource = "api" | "fallback";

export type ExploreModulesPayload = ExploreModulesResponse & {
  source: ExploreModulesSource;
};

const EXPLORE_FALLBACK_FLAGS = {
  explore_breathing: true,
  explore_psychoeducation: true,
  explore_trending: true
} as const;

type ExploreFallbackCopy = {
  modules: ExploreModule[];
};

const EXPLORE_FALLBACK_TEMPLATE: ExploreFallbackCopy = {
  modules: [
    {
      id: "breathing-reset",
      moduleType: "breathing_exercise",
      title: "Breathing reset for today",
      description: "A quick calming practice that steadies your heart rate in about five minutes.",
      cadenceLabel: "4-7-8 breathing cadence",
      durationMinutes: 5,
      steps: [
        {
          label: "Find your posture",
          instruction: "Sit upright, relax your shoulders, gently close your eyes.",
          durationSeconds: 10
        },
        {
          label: "Inhale for 4",
          instruction: "Draw a slow, steady breath through your nose while counting to four.",
          durationSeconds: 16
        },
        {
          label: "Hold for 7",
          instruction: "Keep your shoulders soft and count to seven before releasing.",
          durationSeconds: 28
        },
        {
          label: "Exhale for 8",
          instruction: "Release through your mouth and notice ease returning to your chest.",
          durationSeconds: 32
        }
      ],
      recommendedFrequency: "Repeat two to three rounds before bed or whenever anxiety climbs.",
      ctaLabel: "Begin practice",
      ctaAction: "/app/practices/breathing"
    },
    {
      id: "psychoeducation",
      moduleType: "psychoeducation",
      title: "Self-guided essentials",
      description: "Curated resources that support the most common sleep and anxiety themes.",
      resources: [
        {
          id: "micro-steps",
          title: "Break anxiety into micro steps",
          summary: "Track triggers, practice 4-7-8 breathing, and write a supportive affirmation.",
          readTimeMinutes: 6,
          tags: ["Stress management", "Self-awareness"],
          resourceType: "article"
        },
        {
          id: "sleep-hygiene",
          title: "Sleep rhythm reboot",
          summary: "Set a consistent bedtime routine using gentle wind-down rituals.",
          readTimeMinutes: 5,
          tags: ["Sleep routine", "Relaxation"],
          resourceType: "article"
        },
        {
          id: "body-scan",
          title: "Three-minute body scan audio",
          summary: "Scan for tension and release it with guided breathing cues.",
          readTimeMinutes: 3,
          tags: ["Mindfulness"],
          resourceType: "audio"
        }
      ],
      ctaLabel: "View more",
      ctaAction: "/app/library"
    },
    {
      id: "trending-topics",
      moduleType: "trending_topics",
      title: "Trending focus areas",
      description: "Based on recent conversations, these practices deserve extra attention.",
      topics: [
        {
          name: "Stress management",
          momentum: 70,
          trend: "up",
          summary: "Anxiety mentions are rising; pairing with breathing work eases the spike."
        },
        {
          name: "Sleep rhythm",
          momentum: 58,
          trend: "steady",
          summary: "Keeping a bedtime ritual helps you fall asleep sooner."
        }
      ],
      insights: [
        "A ten-minute wind-down routine can shorten the time it takes to fall asleep.",
        "Jotting triggers and practicing 4-7-8 breathing can steady your heart in under two minutes."
      ],
      ctaLabel: "See practice ideas",
      ctaAction: "/app/trends"
    }
  ]
};

function createFallbackExploreResponse(locale: string): ExploreModulesResponse {
  const modules = JSON.parse(JSON.stringify(EXPLORE_FALLBACK_TEMPLATE.modules)) as ExploreModule[];
  return {
    locale,
    evaluatedFlags: { ...EXPLORE_FALLBACK_FLAGS },
    modules
  };
}

function coerceModuleType(raw: unknown): ExploreModuleType | null {
  const value = typeof raw === "string" ? raw : raw?.toString?.();
  if (value === "breathing_exercise" || value === "psychoeducation" || value === "trending_topics") {
    return value;
  }
  return null;
}

function normalizeBreathingModule(raw: unknown): BreathingModule {
  const data = asRecord(raw) ?? {};
  const steps = asArray(data.steps, (step) => {
    const stepData = asRecord(step);
    if (!stepData) {
      return null;
    }
    return {
      label: asString(stepData.label),
      instruction: asString(stepData.instruction),
      durationSeconds: asNumber(stepData.durationSeconds ?? stepData.duration_seconds)
    };
  });

  return {
    id: asString(data.id, "breathing") || "breathing",
    moduleType: "breathing_exercise",
    title: asString(data.title, "Breathing practice") || "Breathing practice",
    description: asString(data.description),
    featureFlag: asString(data.featureFlag ?? data.feature_flag) || undefined,
    ctaLabel: asString(data.ctaLabel ?? data.cta_label) || undefined,
    ctaAction: asString(data.ctaAction ?? data.cta_action) || undefined,
    durationMinutes: asNumber(data.durationMinutes ?? data.duration_minutes, 5),
    cadenceLabel: asString(data.cadenceLabel ?? data.cadence_label),
    steps,
    recommendedFrequency: asString(data.recommendedFrequency ?? data.recommended_frequency)
  };
}

function normalizePsychoeducationModule(raw: unknown): PsychoeducationModule {
  const data = asRecord(raw) ?? {};
  const resources: PsychoeducationResource[] = asArray(data.resources, (resource) => {
    const resourceData = asRecord(resource);
    if (!resourceData) {
      return null;
    }
    const id =
      asString(resourceData.id) ||
      (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function"
        ? crypto.randomUUID()
        : Math.random().toString(36).slice(2));
    return {
      id,
      title: asString(resourceData.title),
      summary: asString(resourceData.summary),
      readTimeMinutes: asNumber(resourceData.readTimeMinutes ?? resourceData.read_time_minutes),
      tags: asStringArray(resourceData.tags),
      resourceType: asString(resourceData.resourceType ?? resourceData.resource_type, "article") || "article",
      url: resourceData.url === null ? null : asString(resourceData.url) || null
    };
  });

  return {
    id: asString(data.id, "psychoeducation") || "psychoeducation",
    moduleType: "psychoeducation",
    title: asString(data.title, "Psychoeducation") || "Psychoeducation",
    description: asString(data.description),
    featureFlag: asString(data.featureFlag ?? data.feature_flag) || undefined,
    ctaLabel: asString(data.ctaLabel ?? data.cta_label) || undefined,
    ctaAction: asString(data.ctaAction ?? data.cta_action) || undefined,
    resources
  };
}

function normalizeTrendingModule(raw: unknown): TrendingTopicsModule {
  const data = asRecord(raw) ?? {};
  const topics: TrendingTopic[] = asArray(data.topics, (topic) => {
    const topicData = asRecord(topic);
    if (!topicData) {
      return null;
    }
    const trend =
      topicData.trend === "up" || topicData.trend === "down" || topicData.trend === "steady"
        ? (topicData.trend as TrendingTopic["trend"])
        : "steady";
    const name = asString(topicData.name);
    if (!name) {
      return null;
    }
    return {
      name,
      momentum: asNumber(topicData.momentum),
      trend,
      summary: asString(topicData.summary)
    };
  });

  return {
    id: asString(data.id, "trending") || "trending",
    moduleType: "trending_topics",
    title: asString(data.title, "Trending focus areas") || "Trending focus areas",
    description: asString(data.description),
    featureFlag: asString(data.featureFlag ?? data.feature_flag) || undefined,
    ctaLabel: asString(data.ctaLabel ?? data.cta_label) || undefined,
    ctaAction: asString(data.ctaAction ?? data.cta_action) || undefined,
    topics,
    insights: asStringArray(data.insights)
  };
}

function normalizeModule(raw: unknown): ExploreModule | null {
  const data = asRecord(raw);
  const type = coerceModuleType(data?.moduleType ?? data?.module_type);
  if (!type) {
    return null;
  }

  switch (type) {
    case "breathing_exercise":
      return normalizeBreathingModule(data);
    case "psychoeducation":
      return normalizePsychoeducationModule(data);
    case "trending_topics":
      return normalizeTrendingModule(data);
    default:
      return null;
  }
}

function normalizeResponse(raw: unknown, fallbackLocale: string): ExploreModulesResponse {
  const data = asRecord(raw);
  if (!data) {
    return createFallbackExploreResponse(fallbackLocale);
  }

  const modules = asArray(data.modules, (module) => normalizeModule(module)).filter(
    (module): module is ExploreModule => Boolean(module)
  );

  const evaluatedFlags: Record<string, boolean> = {};
  const flagsRecord = asRecord(data.evaluatedFlags);
  if (flagsRecord) {
    for (const [key, value] of Object.entries(flagsRecord)) {
      evaluatedFlags[key] = Boolean(value);
    }
  }

  return {
    locale: asString(data.locale, fallbackLocale) || fallbackLocale,
    modules,
    evaluatedFlags
  };
}

async function requestExploreModules(userId: string, locale: string): Promise<ExploreModulesResponse> {
  const baseUrl = getApiBaseUrl();
  const params = new URLSearchParams();
  params.set("userId", userId);
  if (locale) {
    params.set("locale", locale);
  }

  const response = await fetch(`${baseUrl}/api/explore/modules?${params.toString()}`, {
    credentials: "include",
    headers: withAuthHeaders({
      Accept: "application/json"
    })
  });

  if (!response.ok) {
    throw new Error(`Failed to load explore modules (status ${response.status}).`);
  }

  const payload = await response.json();
  return normalizeResponse(payload, locale);
}

export async function loadExploreModules(userId: string, locale: string): Promise<ExploreModulesPayload> {
  try {
    const response = await requestExploreModules(userId, locale);
    if (response.modules.length === 0) {
      return { ...createFallbackExploreResponse(locale), source: "fallback" };
    }
    return { ...response, source: "api" };
  } catch (error) {
    console.warn("[ExploreModules] Falling back to preset modules:", error);
    return { ...createFallbackExploreResponse(locale), source: "fallback" };
  }
}
